src/model.js

```diff
@@ -2808,14 +2808,26 @@ class Model {
           options.updateOnDuplicate = options.updateOnDuplicate.map(attr => model.rawAttributes[attr].field || attr);
           // Get primary keys for postgres to enable updateOnDuplicate
           options.upsertKeys = _.chain(model.primaryKeys).values().map('field').value();
+
+          let onConflictKeys = [];
           if (Object.keys(model.uniqueKeys).length > 0) {
-            options.upsertKeys = _.chain(model.uniqueKeys)
+            onConflictKeys = _.chain(model.uniqueKeys)
               .values()
               .filter(c => c.fields.length >= 1)
               .map(c => c.fields)
               .reduce(c => c[0])
               .value();
           }
+          model._indexes.forEach(index => {
+            if (index.unique) {
+              index.fields.forEach(uniqueKey => {
+                onConflictKeys.push(uniqueKey);
+              });
+            }
+          });
```

Comment:
Since this always happens we can avoid the middle `let variable` and have it this way

```js
const onConflictKeys = model._indexes.reduce((acc, modelIndex) => {
  if (modelIndex.unique) {
    modelIndex.fields.forEach(uniqueKey => {
      acc.push(uniqueKey);
    });
  }

  return acc;
}, []);

if (Object.keys(model.uniqueKeys).length > 0) {
  const restOnConflictKeys = _.chain(model.uniqueKeys)
    .values()
    .filter((c) => c.fields.length >= 1)
    .map((c) => c.fields)
    .reduce((c) => c[0])
    .value();

  onConflictKeys.push(...restOnConflictKeys);
}

if (onConflictKeys.length > 0) {
  options.upsertKeys = onConflictKeys;
}
```

Reply:
That looks like it would work have you tested it?

Reply:
I have not, but the test case you added should still pass with this and should not break anything else right? All we did here was go through the `model._indexes` first to avoid the middle `let onConflictKeys`.

Reply:
Yep correct.

Reply:
Let me know when you push that change. I'll stamp mine after that.